experimental/packages/opentelemetry-instrumentation/src/instrumentation.ts

```diff
@@ -173,4 +175,33 @@ export abstract class InstrumentationAbstract implements Instrumentation {
     | InstrumentationModuleDefinition
     | InstrumentationModuleDefinition[]
     | void;
+
+  /**
+   * Execute instrumentation event hook, if configured, and log any errors.
+   * Any semantics of the event type and eventInfo are defined by the specific instrumentation.
+   * @param hookHandler The optional hook handler which the user has configured via instrumentation config\
+   * @param eventName The name of the event being triggered, for logging purposes
+   * @param span The span to which the hook should be applied
+   * @param eventInfo The event info to be passed to the hook
+   */
+  protected runInstrumentationEventHook<EventInfoType>(
```

Comment:
Should we allow to specify return type?

Reply:
What will the instrumentation do with the return value? at the moment the hook is just an invocation of whatever logic there is in the hook which can crate side effects on span or somewhere else. returning a value should communicate something to the instrumentation, which we don't do at the moment with current instrumentations. Do you have any suggestion in mind?

My thinking was to keep it simple so that the hooks are easy to use and understand. Any custom hook that need to return some custom value, or receive more arguments, can still define the hook per usage and invoke the hook without the aid of the `runInstrumentationEventHook` function, and take care of checking for nullish hook, exceptions and log error if needed.

I do have some concerns that maybe the name I choose "InstrumentaionEvent" is a bit confusing, as it is only relevant in the context of a span. perhaps we can rename it to "InstrumentationSpanEvent"?

Reply:
I have no common use case in mind for spans. This might be different if instrumentations provide hooks for metrics/logs/...
But in that case we anyway need something new as `span` is a mandatory argument here.